File: core/browseruse_engine/protocol.py
"""
BrowserUse Engine - 统一引擎协议

定义 EngineProtocol 接口，BrowserUseEngine 和 StagehandGoogleEngine 都实现此协议，
实现两个引擎的互换使用。

使用示例:
    ```python
    from core.browseruse_engine.protocol import EngineProtocol

    async def run_task(engine: EngineProtocol, task: str):
        # 可以传入 BrowserUseEngine 或 StagehandGoogleEngine
        result = await engine.run(task)
        return result
    ```
"""

from typing import Protocol, Optional, Any, Dict, List, Type, TypeVar, runtime_checkable
from dataclasses import dataclass, field
from enum import Enum


# ==================== 结果类型定义 ====================

class OperationStatus(Enum):
    """操作状态"""
    SUCCESS = "success"
    FAILED = "failed"
    PARTIAL = "partial"
    TIMEOUT = "timeout"
    BLOCKED = "blocked"


@dataclass
class NavigationResult:
    """导航结果"""
    success: bool
    url: str
    final_url: Optional[str] = None
    error: Optional[str] = None
    duration_ms: float = 0.0

    def __bool__(self) -> bool:
        return self.success


@dataclass
class ActionResult:
    """动作执行结果"""
    success: bool
    message: str = ""
    error: Optional[str] = None
    extracted_content: Optional[str] = None
    duration_ms: float = 0.0

    def __bool__(self) -> bool:
        return self.success


@dataclass
class ExtractResult:
    """数据提取结果"""
    success: bool
    data: Optional[Dict[str, Any]] = None
    error: Optional[str] = None
    duration_ms: float = 0.0

    def __bool__(self) -> bool:
        return self.success


@dataclass
class ObserveResult:
    """页面观察结果"""
    success: bool
    elements: List[Dict[str, Any]] = field(default_factory=list)
    error: Optional[str] = None
    duration_ms: float = 0.0

    def __bool__(self) -> bool:
        return self.success


@dataclass
class AgentStep:
    """Agent 执行步骤记录"""
    step_number: int
    thinking: str = ""
    action_name: str = ""
    action_params: Dict[str, Any] = field(default_factory=dict)
    result: Optional[ActionResult] = None
    browser_url: str = ""
    timestamp: float = 0.0


@dataclass
class AgentResult:
    """Agent 任务执行结果"""
    success: bool
    message: str = ""
    error: Optional[str] = None
    extracted_content: Optional[str] = None
    steps: List[AgentStep] = field(default_factory=list)
    total_steps: int = 0
    duration_ms: float = 0.0

    def __bool__(self) -> bool:
        return self.success


# ==================== 引擎协议定义 ====================

T = TypeVar('T')


@runtime_checkable
class EngineProtocol(Protocol):
    """
    AI 浏览器引擎统一协议

    BrowserUseEngine 和 StagehandGoogleEngine 都应实现此协议，
    使得两个引擎可以互换使用。

    核心方法:
        - navigate: 导航到 URL
        - act: 执行自然语言指令 (单步)
        - extract: 提取页面数据
        - observe: 观察页面元素
        - run: 执行完整 Agent 任务 (多步)

    生命周期:
        - start: 启动引擎
        - stop: 关闭引擎
        - is_initialized: 检查初始化状态
    """

    @property
    def is_initialized(self) -> bool:
        """是否已初始化"""
        ...

    async def start(self) -> None:
        """启动引擎 (本地模式)"""
        ...

    async def stop(self) -> None:
        """关闭引擎"""
        ...

    async def navigate(
        self,
        url: str,
        wait_until: str = "domcontentloaded",
        timeout: float = 30000,
    ) -> NavigationResult:
        """
        导航到指定 URL

        Args:
            url: 目标 URL
            wait_until: 等待条件 ("load", "domcontentloaded", "networkidle")
            timeout: 超时时间（毫秒）

        Returns:
            NavigationResult
        """
        ...

    async def act(
        self,
        instruction: str,
        timeout: float = 30000,
    ) -> ActionResult:
        """
        执行自然语言指令 (单步)

        使用 AI 理解指令并在页面上执行相应操作。

        Args:
            instruction: 自然语言指令 (如 "点击登录按钮")
            timeout: 超时时间（毫秒）

        Returns:
            ActionResult
        """
        ...

    async def extract(
        self,
        instruction: str,
        schema: Optional[Type[T]] = None,
        timeout: float = 30000,
    ) -> ExtractResult:
        """
        提取页面数据

        使用 AI 从页面提取结构化数据。

        Args:
            instruction: 提取描述 (如 "提取所有商品价格")
            schema: Pydantic 模型类，用于结构化输出
            timeout: 超时时间（毫秒）

        Returns:
            ExtractResult
        """
        ...

    async def observe(
        self,
        instruction: str,
        timeout: float = 30000,
    ) -> ObserveResult:
        """
        观察页面元素

        使用 AI 分析页面并找到符合描述的元素。

        Args:
            instruction: 观察描述 (如 "找到所有输入框")
            timeout: 超时时间（毫秒）

        Returns:
            ObserveResult
        """
        ...

    async def run(
        self,
        task: str,
        max_steps: int = 50,
        on_step: Optional[Any] = None,
    ) -> AgentResult:
        """
        执行完整 Agent 任务 (多步)

        使用 Agent 循环自动完成复杂任务。

        Args:
            task: 任务描述
            max_steps: 最大步数
            on_step: 步骤回调函数

        Returns:
            AgentResult
        """
        ...


# ==================== 辅助函数 ====================

def is_engine(obj: Any) -> bool:
    """检查对象是否实现了 EngineProtocol"""
    return isinstance(obj, EngineProtocol)

File: desktop/src/browseruse/protocol.ts
/**
 * BrowserUse Engine - 统一引擎协议（Node 重写）
 * 对标 core/browseruse_engine/protocol.py
 *
 * BrowserUseEngine 与 StagehandGoogleEngine 都实现此协议，两者可互换。
 *
 * 移植约定（与 engine/types.ts 一致）：
 *   - dataclass 字段一律**必需**，默认值由 createXxx() 工厂提供
 *   - 字段名保留 Python 的 snake_case（duration_ms 等），便于逐字对拍
 */

/** 操作状态 —— 对标 protocol.py 的 OperationStatus */
export type OperationStatus = "success" | "failed" | "partial" | "timeout" | "blocked";
export const OperationStatusValues = {
  SUCCESS: "success",
  FAILED: "failed",
  PARTIAL: "partial",
  TIMEOUT: "timeout",
  BLOCKED: "blocked",
} as const;

/** 导航结果 */
export interface NavigationResult {
  success: boolean;
  url: string;
  final_url: string | null;
  error: string | null;
  duration_ms: number;
}

export function createNavigationResult(overrides: Partial<NavigationResult> = {}): NavigationResult {
  const base: NavigationResult = {
    success: false,
    url: "",
    final_url: null,
    error: null,
    duration_ms: 0,
  };
  return applyOverrides(base, overrides);
}

/** 动作执行结果 */
export interface ActionResult {
  success: boolean;
  message: string;
  error: string | null;
  extracted_content: string | null;
  duration_ms: number;
}

export function createActionResult(overrides: Partial<ActionResult> = {}): ActionResult {
  const base: ActionResult = {
    success: false,
    message: "",
    error: null,
    extracted_content: null,
    duration_ms: 0,
  };
  return applyOverrides(base, overrides);
}

/** 数据提取结果 */
export interface ExtractResult<T = Record<string, unknown>> {
  success: boolean;
  data: T | null;
  error: string | null;
  duration_ms: number;
}

export function createExtractResult<T = Record<string, unknown>>(
  overrides: Partial<ExtractResult<T>> = {},
): ExtractResult<T> {
  const base: ExtractResult<T> = {
    success: false,
    data: null,
    error: null,
    duration_ms: 0,
  };
  return applyOverrides(base, overrides);
}

/** 页面观察结果 */
export interface ObserveResult {
  success: boolean;
  elements: Record<string, unknown>[];
  error: string | null;
  duration_ms: number;
}

export function createObserveResult(overrides: Partial<ObserveResult> = {}): ObserveResult {
  const base: ObserveResult = {
    success: false,
    elements: [],
    error: null,
    duration_ms: 0,
  };
  return applyOverrides(base, overrides);
}

/** Agent 执行步骤记录 */
export interface AgentStep {
  step_number: number;
  thinking: string;
  action_name: string;
  action_params: Record<string, unknown>;
  result: ActionResult | null;
  browser_url: string;
  timestamp: number;
}

export function createAgentStep(overrides: Partial<AgentStep> = {}): AgentStep {
  const base: AgentStep = {
    step_number: 0,
    thinking: "",
    action_name: "",
    action_params: {},
    result: null,
    browser_url: "",
    timestamp: 0,
  };
  return applyOverrides(base, overrides);
}

/** Agent 任务执行结果 */
export interface AgentResult {
  success: boolean;
  message: string;
  error: string | null;
  extracted_content: string | null;
  steps: AgentStep[];
  total_steps: number;
  duration_ms: number;
}

export function createAgentResult(overrides: Partial<AgentResult> = {}): AgentResult {
  const base: AgentResult = {
    success: false,
    message: "",
    error: null,
    extracted_content: null,
    steps: [],
    total_steps: 0,
    duration_ms: 0,
  };
  return applyOverrides(base, overrides);
}

/** 覆盖时跳过 undefined，等价于 Python 的「不传即取默认」 */
function applyOverrides<T extends object>(base: T, overrides: Partial<T>): T {
  for (const [k, v] of Object.entries(overrides)) {
    if (v !== undefined) (base as unknown as Record<string, unknown>)[k] = v;
  }
  return base;
}

/** 导航参数（Python 的 wait_until / timeout 关键字参数） */
export interface NavigateOptions {
  waitUntil?: string;
  /** 毫秒；对应 Python 的 timeout */
  timeoutMs?: number;
}

/**
 * AI 浏览器引擎统一协议 —— 对标 protocol.py 的 EngineProtocol。
 *
 * Python 用 runtime_checkable Protocol + isinstance 检查；
 * TS 侧改为结构化接口，一致性在编译期由 `satisfies` 断言保证
 * （见 engine.ts 末尾的 __protocolConformance）。
 */
export interface EngineProtocol {
  readonly isInitialized: boolean;
  start(): Promise<void>;
  stop(): Promise<void>;
  navigate(url: string, options?: NavigateOptions): Promise<NavigationResult>;
  act(instruction: string, timeoutMs?: number): Promise<ActionResult>;
  extract<T = Record<string, unknown>>(
    instruction: string,
    schema?: unknown,
    options?: { timeoutMs?: number; maxSteps?: number },
  ): Promise<ExtractResult<T>>;
  observe(instruction: string, timeoutMs?: number): Promise<ObserveResult>;
  run(
    task: string,
    options?: { maxSteps?: number; onStep?: (step: Record<string, unknown>) => void },
  ): Promise<AgentResult>;
}

/**
 * 检查对象是否实现了 EngineProtocol —— 对标 protocol.py 的 is_engine()。
 * Python 靠 runtime_checkable 只校验方法存在性，这里做同样的鸭子检查。
 */
export function isEngine(obj: unknown): boolean {
  if (obj === null || typeof obj !== "object") return false;
  const o = obj as Record<string, unknown>;
  const methods = ["start", "stop", "navigate", "act", "extract", "observe", "run"];
  return methods.every((m) => typeof o[m] === "function");
}
